### PR TITLE
Bp 397 rest routes

### PR DESCRIPTION
**Pull request by MaximilianV**
Contains some improvements and refactorings
Now the basic backend mock data flow is implemented and the grid view calls are working fine.

### CODE REVIEW

File: src/Utils/ConnectionComponent.js [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+		} else if (promiseState.rejected) {
+			return <ErrorMessage message={promiseState.reason.message}/>;
+		}
+		return null;
+	}
+	
+	static switchFetch(input, init) {
+		if(Config.useBackendMock) {
+			const promise = new Promise(resolve => resolve(BackendMock.handleRequest(input, init)));
+			console.log(promise);
```
Comment (hpibptlab) [inline src/Utils/ConnectionComponent.js]: MAJOR Remove this logging statement. [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3AS2228)

File: src/Utils/MockData/BackendMock.js [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+		return Hierarchy;
+	}
+	
+	static getEntity(params) {
+		return Entity.find((entity) => {
+			return entity.Id = params[1];
+		});
+	}
+	
+	static getEntityTypeAttributes(params) {
```
Comment (hpibptlab) [inline src/Utils/MockData/BackendMock.js]: MAJOR Remove the unused function parameter "params". [rule](http://bpt-lab.org/sonarqube/coding_rules#rule_key=javascript%3AUnusedFunctionArgument)

**Comment by hpibptlab** [comment]
SonarQube analysis reported 2 issues
* ![MAJOR][MAJOR] 2 major

Watch the comments in this conversation to review them.

[MAJOR]: https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR'

File: src/GridView/CardGrid.js [hunk 3]
```diff
@@ -1,53 +1,53 @@
...
+		const connectionIncomplete = super.render(this.props.entities);
+		if(connectionIncomplete) {
+			return connectionIncomplete;
+		}
+		return (
+			<div className="card-grid d-flex">
+				{this.props.entities.value.map((childEntity, index) => {
+					return (
+						<Card className="card" key={index}>
+							<CardTitle title={childEntity.Name} subtitle={this.props.entityType.name}/>
```
Comment (friedow) [inline src/GridView/CardGrid.js]: Break this line.

File: src/GridView/CardGrid.js [hunk 4]
```diff
@@ -1,53 +1,53 @@
...
 		}
+		const connectionIncomplete = super.render(this.props.entities);
+		if(connectionIncomplete) {
+			return connectionIncomplete;
+		}
+		return (
+			<div className="card-grid d-flex">
+				{this.props.entities.value.map((childEntity, index) => {
+					return (
+						<Card className="card" key={index}>
```
Comment (friedow) [inline src/GridView/CardGrid.js]: Break this line.

File: src/GridView/CardGrid.js [hunk 5]
```diff
@@ -1,53 +1,53 @@
...
+							<CardTitle title={childEntity.Name} subtitle={this.props.entityType.name}/>
+							<CardText>
+								Lorem ipsum dolor sit amet, consectetur adipiscing elit.
+								Donec mattis pretium massa. Aliquam erat volutpat. Nulla facilisi.
+								Donec vulputate interdum sollicitudin. Nunc lacinia auctor quam sed pellentesque.
+								Aliquam dui mauris, mattis quis lacus id, pellentesque lobortis odio.
+							</CardText>
+							<CardActions className="d-flex">
+								<FlatButton label="Children" className="card-button" href={`./${childEntity.Id}`}/>
```
Comment (friedow) [inline src/GridView/CardGrid.js]: Break this line.

File: src/GridView/CardGrid.js [hunk 6]
```diff
@@ -1,53 +1,53 @@
...
+								<FlatButton label="Inspect" className="card-button" href={`/details/${childEntity.Id}`}/>
```
Comment (friedow) [inline src/GridView/CardGrid.js]: Break this line.

File: src/GridView/CardGrid.js (hunk 5 above)
Comment (friedow) [reply src/GridView/CardGrid.js]: This link is probably wrong. it should be /grid/${childEntity.Id}

File: src/GridView/GridView.js [hunk 7]
```diff
@@ -17,38 +15,43 @@ class GridView extends Component {
 				return entityType.ParentId === parentEntityTypeId;
 			}));
 		});
+		return childEntities;
 	}
 	
 	render() {
-		const { hierarchy, entity } = this.props;
+		let { hierarchy, entity } = this.props;
```
Comment (friedow) [inline src/GridView/GridView.js]: I would prefer an explicit assignment.

Comment (friedow) [inline src/GridView/GridView.js]: Can you extract Line 22 - 26 into a method?

File: src/Utils/ErrorMessage.js [hunk 8]
```diff
@@ -11,7 +11,7 @@ class ErrorMessage extends Component {
 		return (
 			<div className={css(AppStyles.dFlex, AppStyles.alignItemsCenter, AppStyles.flexDirectionColumn)}>
 				<IconWarning className={css(DefinitionStyles.errorColor, ErrorMessageStyles.iconSize)}/>
-				<p dangerouslySetInnerHTML={{__html: this.props.message}} />
+				<p>{this.props.message}</p>
```
Comment (friedow) [inline src/Utils/ErrorMessage.js:14]: Now this way again? (Merge failure?)

File: src/Utils/MockData/BackendMock.js [hunk 9]
```diff
@@ -0,0 +1,51 @@
+import Config from './../../config/config.js';
+import Hierarchy from './Transport for London/Hierarchy.js';
```
Comment (friedow) [inline src/Utils/MockData/BackendMock.js]: We should really avoid directory and filenames with whitespaces.

File: src/Utils/MockData/BackendMock.js [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+			.set(/^entitytype\/hierarchy$/i, BackendMock.getEntityTypeHierarchy)
+			.set(/^entitytype\/(-?\d+)\/attributes$/i, BackendMock.getEntityTypeAttributes)
+			.set(/^entity\/(-?\d+)\/children\/type\/(-?\d+)\/(((\w)+)\+)*(\w)*$/i, BackendMock.getChildEntitiesOfEntityType)
+			.set(/^entity\/(-?\d+)$/i, BackendMock.getEntity);
+	}
+		
+	static handleRequest(request) {
+		const RouteMapper = BackendMock.getRouteMapper();
+		const cleanedRoute = request.url.replace(Config.backendRESTRoute,"");
+		const targetRoute = [...RouteMapper.keys()].find((route) => {
```
Comment (friedow) [inline src/Utils/MockData/BackendMock.js:19]: What does `...` do?

File: src/Utils/MockData/BackendMock.js [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+		});
+	}
+	
+	static getEntityTypeAttributes(params) {
+		return EntityType[params[1]]["attributes"];
+	}
+	
+	static getChildEntitiesOfEntityType(params) {
+		return Entity.filter((Entity) => {
+			return (Entity.ParentId === parseInt(params[1], 10) && Entity.TypeId === parseInt(params[2], 10));
```
Comment (friedow) [inline src/Utils/MockData/BackendMock.js]: Thats a long line.

File: src/index.js [hunk 12]
```diff
@@ -1,14 +1,17 @@
...
-import {BrowserRouter as Router, Route} from 'react-router-dom';
+import {BrowserRouter as Router, Route, Redirect} from 'react-router-dom';
 import GridView from './GridView/GridView';
 import DetailView from './DetailView/DetailView';
 import EventTypesView from './EventTypesView/EventTypesView';
 
 ReactDOM.render(
 	<Router>
 		<App>
+			<Route exact path="/" render={() => {
```
Comment (friedow) [inline src/index.js:12]: Please merge carefully. Tom solved this too...

**Review by friedow — CHANGES_REQUESTED** [review]

File: src/Utils/MockData/BackendMock.js (hunk 10 above)
Comment (MaximilianV) [reply src/Utils/MockData/BackendMock.js:19]: `.keys()` gives you an iterator for _RouteMapper_. But `find()` expects an array. The `...` operator _converts_ the iterator to an array we can operate on.

File: src/Utils/ErrorMessage.js (hunk 8 above)
Comment (MaximilianV) [reply src/Utils/ErrorMessage.js:14]: Yes we could avoid the `dangerouslySetInnerHTML` 😄 
(Not a merge failure)

File: src/index.js (hunk 12 above)
Comment (MaximilianV) [reply src/index.js:12]: @tommartensen approves this routing 👍

**Review by friedow — APPROVED** [review]

**Pull request merged**